src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs

```diff
@@ -79,19 +83,53 @@ public static Completion.CompletionTrigger LSPToRoslynCompletionTrigger(LSP.Comp
             }
             else if (context.TriggerKind == LSP.CompletionTriggerKind.Invoked)
             {
-                return Completion.CompletionTrigger.Invoke;
+                if (context is not LSP.VSCompletionContext vsCompletionContext)
+                {
+                    return Completion.CompletionTrigger.Invoke;
+                }
+
+                switch (vsCompletionContext.InvokeKind)
+                {
+                    case LSP.VSCompletionInvokeKind.Explicit:
+                        return Completion.CompletionTrigger.Invoke;
+
+                    case LSP.VSCompletionInvokeKind.Typing:
+                        var insertionChar = await GetInsertionCharacter(document, position, cancellationToken).ConfigureAwait(false);
```

Comment:
Super interesting that Roslyn cares about the insertionChar. Out of sheer curiosity what sort of decisions do you make based off this character that isn't already covered by a "triggered" completion?

Reply:
Sometime we need to make nuanced decision like whether to trigger completion and whether soft-selection should be used, even we have already know the trigger kind. You can see those cases where `CompletionTrigger.Character` is used
[http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Completion/CompletionTrigger.cs,16a654be86e83be7,references](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Completion/CompletionTrigger.cs,16a654be86e83be7,references)

---
In reply to: [570448532](https://github.com/dotnet/roslyn/pull/51002#discussion_r570448532) [](ancestors = 570448532)

Reply:
Hmm, looks like it's only the DateTime completion handling that really cares because the other pieces that care just do it for deletion which is handled separately. And even at that it's not really doing an identifier check, learning a good bit, thanks for the code links.